sklearn/feature_selection/_rfe.py

```diff
@@ -42,6 +42,20 @@ def _rfe_single_fit(rfe, estimator, X, y, train, test, scorer):
     ).scores_
 
 
+def _estimator_has(attr):
+    """Check if the provided estimator has the attribute.
+
+    We check if the attribute is available in the fitted estimator first
+    (i.e. `self.estimator_`). If not available, we check the estimator
+    passed at initialization (i.e. `self.estimator`).
+    """
+    return lambda self: (
+        hasattr(self.estimator_, attr)
+        if hasattr(self, "estimator_")
+        else hasattr(self.estimator, attr)
+    )
```

Comment:
Do we really want to copy and paste this in several modules? I think we should factorize the most common predicates (`_estimator_has`, `_base_estimator_has`, and maybe others in a `sklearn.utils.metaestimators` next to the `available_if` method itself).

We can keep those predicates private for now.

WDYT @jnothman?

Reply:
Hi, hope you don't mind me chiming in here. 

I took a quick look at all the applicable estimators (except multi-class and pipeline). It seems like most checks should follow a similar structure (i.e. check fitted estimator first if available and fall back to unfitted estimator) with slightly different estimators names.

-  ./sklearn/model_selection/_search.py: `best_estimator_`, `estimator`
-  ./sklearn/feature_selection/_from_model.py: `estimator_,` `estimator`
-  ./sklearn/feature_selection/_rfe.py: `estimator_`, `estimator`
-  ./sklearn/semi_supervised/_self_training.py: `base_estimator_`, `base_estimator`
-  ./sklearn/ensemble/_stacking.py: `final_estimator_`, `final_estimator` (needs more discussion)
-  ./sklearn/ensemble/_bagging.py: `estimators_[0]`, `base_estimator` (needs more discussion)
-  ./examples/cluster/plot_inductive_clustering.py: `classifier_`, `classifier`
-  [optional?]./sklearn/utils/tests/test_testing.py: `delegate`

So maybe we could just have the following function in `metaestimators.py` (with a better function name), and do e.g. `available_if(_estimator_has(["estimator_", "estimator"], "predict"))`

```
def _estimator_has(estimator_names, attr):
    
    def check(self):   # borrowed from _IffHasAttrDescriptor._check
        estimator = None
        for estimator_name in estimator_names:
            try:
                estimator = attrgetter(estimator_name)(self)
                break
            except AttributeError:
                continue

        return estimator is not None and hasattr(estimator, attr)
    
    return check
```

Reply:
@ogrisel We already made some changes where @jnothman was in favour of not factorizing the check.
I would say that as a first step, we could make all the checks local (at least this is easily readable at the cost of duplication). As a second step, we could think about factorizing once all the `available_if` are in the library.
We can see what is best at this stage.